<a href="https://colab.research.google.com/github/rafi1624/Machine-Learning_18_Rafi-Abyantara/blob/main/JS05/JS%2005-Tugas%20Lab%202.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv
2. Fitur CountVectorizer dengan mengaktifkan stop_words
3. Evaluasi hasilnya

In [10]:
import pandas as pd


df = pd.read_csv('/content/spam.csv', encoding='latin-1')
df = df.iloc[:, :2]
df.columns = ['target', 'text']

display(df.head())

,target,text
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [11]:
from sklearn.feature_extraction.text import CountVectorizer


vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(df['text'])
y = df['target']

print(f"Bentuk matrix fitur X: {X.shape}")
print(f"jumlah kata unik: {len(vectorizer.get_feature_names_out())}")

Bentuk matrix fitur X: (5572, 8404)
jumlah kata unik: 8404


In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(f"Bentuk X_train: {X_train.shape}")
print(f"Bentuk X_test: {X_test.shape}")
print(f"Bentuk y_train: {y_train.shape}")
print(f"Bentuk y_test: {y_test.shape}")

Bentuk X_train: (4457, 8404)
Bentuk X_test: (1115, 8404)
Bentuk y_train: (4457,)
Bentuk y_test: (1115,)


In [13]:
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB()
model.fit(X_train, y_train)

print("Model training selsai.")

Model training selsai.


In [14]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

y_pred = model.predict(X_test)

# Evaluasi
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

print(f"\nAkurasi: {accuracy:.4f}")
print("\nClassification Report:\n", report)
print("\nConfusion Matrix:\n", conf_matrix)


Akurasi: 0.9803

Classification Report:
               precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.93      0.93      0.93       150

    accuracy                           0.98      1115
   macro avg       0.96      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115


Confusion Matrix:
 [[954  11]
 [ 11 139]]


Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv
2. Fitur TF-IDF dengan mengaktifkan stop_words
3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer


tfidf_vectorizer = TfidfVectorizer(stop_words='english')

X_tfidf = tfidf_vectorizer.fit_transform(df['text'])
y_tfidf = df['target']

print(f"Bentuk matrix fitur X_tfidf: {X_tfidf.shape}")
print(f"jumlah kata unik (TF-IDF): {len(tfidf_vectorizer.get_feature_names_out())}")

Bentuk matrix fitur X_tfidf: (5572, 8404)
jumlah kata unik (TF-IDF): 8404


In [16]:
from sklearn.model_selection import train_test_split


X_train_tfidf, X_test_tfidf, y_train_tfidf, y_test_tfidf = train_test_split(X_tfidf, y_tfidf, test_size=0.2, random_state=42)

print(f"Bentuk X_train_tfidf: {X_train_tfidf.shape}")
print(f"Bentuk X_test_tfidf: {X_test_tfidf.shape}")
print(f"Bentuk y_train_tfidf: {y_train_tfidf.shape}")
print(f"Bentuk y_test_tfidf: {y_test_tfidf.shape}")

Bentuk X_train_tfidf: (4457, 8404)
Bentuk X_test_tfidf: (1115, 8404)
Bentuk y_train_tfidf: (4457,)
Bentuk y_test_tfidf: (1115,)


In [17]:
from sklearn.naive_bayes import MultinomialNB


model_tfidf = MultinomialNB()
model_tfidf.fit(X_train_tfidf, y_train_tfidf)

print("Model TF-IDF training selesai.")

Model TF-IDF training selesai.


In [18]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

y_pred_tfidf = model_tfidf.predict(X_test_tfidf)

accuracy_tfidf = accuracy_score(y_test_tfidf, y_pred_tfidf)
report_tfidf = classification_report(y_test_tfidf, y_pred_tfidf)
conf_matrix_tfidf = confusion_matrix(y_test_tfidf, y_pred_tfidf)

print(f"\nAkurasi (TF-IDF): {accuracy_tfidf:.4f}")
print("\nClassification Report (TF-IDF):\n", report_tfidf)
print("\nConfusion Matrix (TF-IDF):\n", conf_matrix_tfidf)


Akurasi (TF-IDF): 0.9686

Classification Report (TF-IDF):
               precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.77      0.87       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.93      1115
weighted avg       0.97      0.97      0.97      1115


Confusion Matrix (TF-IDF):
 [[965   0]
 [ 35 115]]


3.
Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

- Hasil NO 2
  Akurasi : 0.9803
- Hasil NO 3
  Akurasi : 0.9686

Secara keseluruhan, model menunjukkan kinerja yang sangat baik dalam mengklasifikasikan pesan spam.

  

4. Berikan kesimpulan fitur mana yang terbaik pada kasus data

Berdasarkan akurasi keseluruhan dan F1-score untuk kelas 'spam', CountVectorizer terbukti menjadi fitur ekstraksi teks yang lebih baik untuk kasus data spam.csv ini. Ini menunjukkan bahwa frekuensi kemunculan kata (seperti yang ditangkap oleh CountVectorizer) lebih informatif dalam membedakan antara spam dan ham untuk dataset ini, bahkan setelah menghapus stop words.